README.md
ajaxmap
=======

Ajax Map is an extension for the TYPO3 CMS. It allows to show a map with locations. 
Display can be filtered interactivly by type, region and categories. 
Categories are shown and selected in a tree view. 
The map view is highly configurable. Regions are defined by KML files. They are displayed as map overlays an can be switched on/off in frontend.  
See also: <https://forge.typo3.org/projects/extension-ajaxmap>

[![Coverage Status](https://coveralls.io/repos/dwenzel/ajaxmap/badge.png)](https://coveralls.io/r/dwenzel/ajaxmap)

## Requirements

* TYPO3 CMS 9

## Installation

* Install extension:
```bash
composer require dwenzel/ajaxmap
``` 

* Include static TypoScript Template 


ChangeLog
2019-08-19         [RELEASE] Release of ajaxmap 2.0.0 (Dirk Wenzel)
2019-08-19 f0136c0 [TASK] marker are now clustered (Dirk Wenzel)
2019-08-19 5ffa25d [TASK] marker clusterer added (WIP) (Dirk Wenzel)
2019-08-19 d5f56e9 [TASK] use fontawesome as default style - obsolete libraries removed (Dirk Wenzel)
2019-08-19 5a5936b [TASK] deprecated fancytree option replaced (Dirk Wenzel)
2019-08-15 9b3cfa3 [TASK] javaScript libs updated - jQuery updated to 3.4.1 - jQuery UI no longer necessary - fancytree updated to 2.31.0 (Dirk Wenzel)

2019-06-05         [RELEASE] Release of ajaxmap 1.0.2 (Dirk Wenzel)
2019-06-05 83cd375 [BUGFIX] extension status corrected (Dirk Wenzel)

2019-06-05         [RELEASE] Release of ajaxmap 1.0.1 (Dirk Wenzel)
2019-05-27 ac38165 [BUGFIX] fixed error where location type tree was not initialized correctly (Dirk Wenzel)
2019-05-27 9add380 [TASK] caching for ChildrenService implemented, - fixed error in TreeUtility where recursion stopped too early - (Dirk Wenzel)
2019-05-27 0f41247 [TASK] dependency from cpsit/geo-location-service updated (Dirk Wenzel)
2019-05-27 2acb685 [TASK] use GeoCodableInterface (Dirk Wenzel)
2019-05-27 9e6e101 [TASK] Place now returns GeoCoordinates from Address (Dirk Wenzel)
2019-05-27 3b87b0c [TASK] use Address implementation from tt_address (Dirk Wenzel)
2019-05-27 34ae7ea [TASK] deprecated annotation replaced (Dirk Wenzel)
2019-05-16 2a59e52 [TASK] mapping for PlaceGroup and LocationType added (Dirk Wenzel)
2019-05-15 1124556 [TASK] code style (Dirk Wenzel)
2019-05-15 9312b4b [TASK] use jquery-2.2.4 (Dirk Wenzel)
2019-05-15 4b14f2d [TASK] RequestMiddlewares simplified (Dirk Wenzel)
2019-05-14 c0a9172 [BUGFIX] fixed error caused by invalid constructor argument count. (Dirk Wenzel)
2019-05-14 eef5d3f [TASK] unit test for AddressDataProvider and CategoryDataProvider added (Dirk Wenzel)
2019-05-14 2cb520e [BUGFIX] typo in table configuration corrected (Dirk Wenzel)
2019-05-14 c8bba9d [TASK] ChildrenService refactored using Doctrine/Dbal (Dirk Wenzel)
2019-05-14 dbcab64 [TASK] require typo3/cms-core:^9.5 (Dirk Wenzel)
2019-05-13 9e71d3a [BUGFIX] failing unit test for MapController fixed (Dirk Wenzel)
2019-05-13 f05dcd8 [TASK] EidDispatcher replaced by Middleware/FrontendJsonApiHandler (Dirk Wenzel)
2019-05-12 c255acf [BUGFIX] test bootstrap path corrected (Dirk Wenzel)
2019-05-12 46cdc4d [BUGFIX] dev requirements updated (Dirk Wenzel)
2019-05-12 c16cca2 [TASK] AjaxController implemented (WIP) (Dirk Wenzel)
2019-05-10 169da2c [TASK] record icons registered (Dirk Wenzel)
2019-04-15 a6d5bec [TASK][CI] prepare unit tests - require-dev nimut/testing-framework - build matrix fixed (Dirk Wenzel)
2019-04-12 bf0d471 [TASK][CI] build matrix updated -  TYPO3 versions before 8.7 are not supported anymore (Dirk Wenzel)
2019-04-12 e564a44  [TASK] vendor name changed (Dirk Wenzel)
2019-04-12 a71e635 [BUGFIX] invalid variabel $_EXTKEY replaced by string (Dirk Wenzel)
2019-04-12 2a077f6 [TASK] deprecated annotations and methods replaced (Dirk Wenzel)
2019-04-12 42ba8bd [TASK] table configuration updated for TYPO3 version 9 (Dirk Wenzel)
2019-04-12 38c5b97 [TASK] require friendsoftypo3/tt-address (Dirk Wenzel)
2019-04-12 120ce44 [TASK] require typo/cms-core instead of typo3/cms (Dirk Wenzel)
2019-04-12 6fa04fc [TASK] require typo3/cms:>=8.4,<=9.5 (Dirk Wenzel)

2019-06-05         [RELEASE] Release of ajaxmap 1.0.0 (Dirk Wenzel)
2019-04-12 938f107 [TASK] composer vendor name changed (Dirk Wenzel)
2016-03-06 d39434e [TASK] CategoryAwareDemandInterface and CategoryAwareDemandTrait implemented. PlaceDemand refactored. (Dirk Wenzel)
2016-03-06 ab227cb [BUGFIX] missing coverage annotation added. (Dirk Wenzel)
2016-03-06 0f5920e [TASK] CategorizableInterface and CategorizableTrait implemented. Map and Place refactored using it. (Dirk Wenzel)
2016-03-06 93957a0 [BUGFIX] missing class in ext_tables fixed. (Dirk Wenzel)
2016-03-06 662f24b  [BUGFIX] restored methods previously removed by mistake (Dirk Wenzel)
2016-03-06 7ea6400 [BUGFIX] match implementation and declaration of SerializableInterface->toArray (Dirk Wenzel)
2016-03-06 74a2b5b [TASK] AbstractEntity removed. All domain objects do now extend AbstractEntity from Extbase framework and implement SerializableInterface (Dirk Wenzel)
2016-03-06 e08fb9b [BUGFIX] obsolete tests for Category removed, failing test fixed. (Dirk Wenzel)
2016-03-06 9638292 [BUGFIX] make declaration of ToJsonTrait->toArray compatible with implementation in ToArrayTrait (Dirk Wenzel)
2016-03-06 3d65e7b [TASK] ToArrayTrait and ToJsonTrait implemented. Category now extends Extbase Category (Dirk Wenzel)
2016-03-05 4a0e1b3 [TASK] fixed broken configuration for coveralls (Dirk Wenzel)
2016-03-05 c037bdd [TASK] Methods of TreeUtility are now non-static. (Dirk Wenzel)
2016-03-05 15fa17e [TASK] ajaxmap now requires php > 5.5 and supports TYPO3 version 6.2 - 7.x (Dirk Wenzel)
2016-01-04 cc7a65f [TASK] get filters for update from partial Map/Data (Dirk Wenzel)
2015-11-03 33fa3ba [TASK] update layers after filtering filters too (Dirk Wenzel)
2015-10-20 ff3d940 [TASK] static layers from model Map are now added to Google Map (Dirk Wenzel)
2015-10-20 9408188 [TASK] field staticLayers added to Map. UnitTests for Map added. (Dirk Wenzel)
2015-10-19 d8afdc6 [TASK] new properties selected and unselectable added to Region model. (Dirk Wenzel)
2015-10-19 4982fdd [TASK] made places tree configurable through partial Map/Data. Allow opening the info window on select of a places tree node. (Dirk Wenzel)
2015-10-17 4fbfe96 [BUGFIX] fixed tests failing under PHP 5.3 caused by short array notation (Dirk Wenzel)
2015-10-16 1c6cb57 [TASK] unit tests for TreeUtility added (Dirk Wenzel)
2015-10-16 ab572d4 [TASK] unit tests for TreeUtility added, use SerializableInterface instead of AbstractEntity, annotation added (Dirk Wenzel)
2015-10-16 33ea61c [BUGFIX] attribute name corrected (Dirk Wenzel)

Configuration/TCA/Place.php
<?php
if (!defined ('TYPO3_MODE')) {
	die ('Access denied.');
}

$TCA['tx_ajaxmap_domain_model_place'] = array(
	'ctrl' => $TCA['tx_ajaxmap_domain_model_place']['ctrl'],
	'interface' => array(
		'showRecordFieldList' => 'sys_language_uid, l10n_parent, l10n_diffsource, hidden, title, geo_coordinates, description, info, place_groups, location_type, regions, content, address',
	),
	'types' => array(
		'1' => array('showitem' => 'sys_language_uid;;;;1-1-1, l10n_parent, l10n_diffsource, hidden;;1, title, geo_coordinates, description, info, place_groups, location_type, regions, content, address,--div--;LLL:EXT:cms/locallang_ttc.xml:tabs.access,starttime, endtime'),
	),
	'palettes' => array(
		'1' => array('showitem' => ''),
	),
	'columns' => array(
		'sys_language_uid' => array(
			'exclude' => 1,
			'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.language',
			'config' => array(
				'type' => 'select',
				'foreign_table' => 'sys_language',
				'foreign_table_where' => 'ORDER BY sys_language.title',
				'items' => array(
					array('LLL:EXT:lang/locallang_general.xml:LGL.allLanguages', -1),
					array('LLL:EXT:lang/locallang_general.xml:LGL.default_value', 0)
				),
			),
		),
		'l10n_parent' => array(
			'displayCond' => 'FIELD:sys_language_uid:>:0',
			'exclude' => 1,
			'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.l18n_parent',
			'config' => array(
				'type' => 'select',
				'items' => array(
					array('', 0),
				),
				'foreign_table' => 'tx_ajaxmap_domain_model_place',
				'foreign_table_where' => 'AND tx_ajaxmap_domain_model_place.pid=###CURRENT_PID### AND tx_ajaxmap_domain_model_place.sys_language_uid IN (-1,0)',
			),
		),
		'l10n_diffsource' => array(
			'config' => array(
				'type' => 'passthrough',
			),
		),
		't3ver_label' => array(
			'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.versionLabel',
			'config' => array(
				'type' => 'input',
				'size' => 30,
				'max' => 255,
			)
		),
		'hidden' => array(
			'exclude' => 1,
			'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.hidden',
			'config' => array(
				'type' => 'check',
			),
		),
		'starttime' => array(
			'exclude' => 1,
			'l10n_mode' => 'mergeIfNotBlank',
			'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.starttime',
			'config' => array(
				'type' => 'input',
				'size' => 13,
				'max' => 20,
				'eval' => 'datetime',
				'checkbox' => 0,
				'default' => 0,
				'range' => array(
					'lower' => mktime(0, 0, 0, date('m'), date('d'), date('Y'))
				),
			),
		),
		'endtime' => array(
			'exclude' => 1,
			'l10n_mode' => 'mergeIfNotBlank',
			'label' => 'LLL:EXT:lang/locallang_general.xml:LGL.endtime',
			'config' => array(
				'type' => 'input',
				'size' => 13,
				'max' => 20,
				'eval' => 'datetime',
				'checkbox' => 0,
				'default' => 0,
				'range' => array(
					'lower' => mktime(0, 0, 0, date('m'), date('d'), date('Y'))
				),
			),
		),
		'title' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.title',
			'config' => array(
				'type' => 'input',
				'size' => 30,
				'eval' => 'trim,required'
			),
		),
		'geo_coordinates' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.geo_coordinates',
			'config' => array(
				'type' => 'input',
				'size' => 30,
				'eval' => 'trim'
			),
		),
		'description' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.description',
			'config' => array(
				'type' => 'text',
				'cols' => 40,
				'rows' => 15,
				'eval' => 'trim'
			),
		),
		'info' => array(
			'exclude' => 1,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.info',
			'config' => array(
				'type' => 'text',
				'cols' => 40,
				'rows' => 15,
				'eval' => 'trim',
				'wizards' => array(
					'RTE' => array(
						'icon' => 'wizard_rte2.gif',
						'notNewRecords'=> 1,
						'RTEonly' => 1,
						'script' => 'wizard_rte.php',
						'title' => 'LLL:EXT:cms/locallang_ttc.xml:bodytext.W.RTE',
						'type' => 'script'
					)
				)
			),
			'defaultExtras' => 'richtext[]',
		),
		'place_groups' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.placeGroups',
			'config' => array(
				'type' => 'select',
				'foreign_table' => 'tx_ajaxmap_domain_model_placegroup',
				'MM' => 'tx_ajaxmap_place_placegroup_mm',
				'renderMode' => 'tree',
				'treeConfig' => array(
					'appearance' => array(
						'expandAll' => 1,
						'maxLevels' => 99,
						'showHeader' => 1
					),
					'parentField' => 'parent',
				),
				'size' => 10,
				'autoSizeMax' => 30,
				'maxitems' => 9999,
				'multiple' => 0,
				'wizards' => array(
					'_PADDING' => 1,
					'_VERTICAL' => 1,
					'edit' => array(
						'type' => 'popup',
						'title' => 'Edit',
						'script' => 'wizard_edit.php',
						'icon' => 'edit2.gif',
						'popup_onlyOpenIfSelected' => 1,
						'JSopenParams' => 'height=350,width=580,status=0,menubar=0,scrollbars=1',
						),
					'add' => Array(
						'type' => 'script',
						'title' => 'Create new',
						'icon' => 'add.gif',
						'params' => array(
							'table' => 'tx_ajaxmap_domain_model_placegroup',
							'pid' => '###CURRENT_PID###',
							'setValue' => 'prepend'
							),
						'script' => 'wizard_add.php',
					),
				),
			),
		),
		'location_type' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.location_type',
			'config' => array(
				'type' => 'select',
				'foreign_table' => 'tx_ajaxmap_domain_model_locationtype',
				'minitems' => 0,
				'maxitems' => 1,
			),
		),
		'icon' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.icon',
			'config' => array(
				'type' => 'group',
				'internal_type' => 'file_reference',
				'show_thumbs' => 1,
				'size' => 1,
				'allowed' => $GLOBALS['TYPO3_CONF_VARS']['GFX']['imagefile_ext'],
				'disallowed' => '',
			),
		),
		'regions' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.regions',
			'config' => array(
				'type' => 'select',
				'foreign_table' => 'tx_ajaxmap_domain_model_region',
				'MM' => 'tx_ajaxmap_place_region_mm',
				'size' => 10,
				'autoSizeMax' => 30,
				'maxitems' => 9999,
				'multiple' => 0,
				'wizards' => array(
					'_PADDING' => 1,
					'_VERTICAL' => 1,
					'add' => Array(
						'type' => 'script',
						'title' => 'Create new',
						'icon' => 'add.gif',
						'params' => array(
							'table' => 'tx_ajaxmap_domain_model_region',
							'pid' => '###CURRENT_PID###',
							'setValue' => 'prepend'
							),
						'script' => 'wizard_add.php',
					),
					'suggest' => array(    
					  	'type' => 'suggest',
					),
				),
			),
		),
		'content' => array(
			'exclude' => 1,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.content',
			'config' => array(
				'type' => 'inline',
				'allowed' => 'tt_content',
				'foreign_table' => 'tt_content',
				//'foreign_field' => 'place',
				'minitems' => 0,
                'maxitems'      => 10,
				'appearance' => array(
                    'collapseAll' => 1,
                    'expandSingle' => 1,
                    'levelLinksPosition' => 'bottom',
                    'useSortable' => 1,
                    'showPossibleLocalizationRecords' => 1,
                    'showRemovedLocalizationRecords' => 1,
                    'showAllLocalizationLink' => 1,
                    'showSynchronizationLink' => 1,
                    'enabledControls' => array(
                        'info' => FALSE,
                    )
                )
			),
		),
		'address' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.address',
			'config' => array(
				'type' => 'group',
				'internal_type' => 'db',
				'allowed' => 'tt_address',
				'foreign_table' => 'tt_address',
				'size' => 1,
				'minitems' => 0,
				'maxitems' => 1,
			),
		),
	),
);

$TCA['tx_ajaxmap_domain_model_place']['columns']['description']['config'] = array(
	'type' => 'text',
	'cols' => 32,
	'rows' => 5,
	'eval' => 'trim'
);

$TCA['tx_ajaxmap_domain_model_place']['interface'] = array(
		'showRecordFieldList' => 'sys_language_uid, l10n_parent, l10n_diffsource, hidden, location_type, title, address, description,  geo_coordinates, place_groups, regions, info, icon, content',
	);
$TCA['tx_ajaxmap_domain_model_place']['types'] = array(
		'1' => array('showitem' => 'sys_language_uid;;;;1-1-1, l10n_parent, l10n_diffsource, hidden;;1, location_type, title, address, description, geo_coordinates,--div--;LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.attribution, place_groups, regions,--div--;LLL:EXT:ajaxmap/Resources/Private/Language/locallang_db.xml:tx_ajaxmap_domain_model_place.content, info, icon, content, --div--;LLL:EXT:cms/locallang_ttc.xml:tabs.access,starttime, endtime'),
);
$TCA['tx_ajaxmap_domain_model_place']['palettes'] = array(
		'1' => array(
			'showitem' => '',
			''
		)
	);
?>
